### PR TITLE
Register custom row constructor function for Gluten.

### PR DESCRIPTION
**Pull request by lviiii**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

Passed UT, TPCH, TPCDS

### CODE REVIEW

File: cpp/velox/compute/RowConstructor.cc [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        });
+      }
+    }
+    RowVectorPtr localResult = std::make_shared<RowVector>(
+        context.pool(),
+        outputType,
+        nulls,
+        rows.size(),
+        std::move(argsCopy),
+        rows.size() /*nullCount*/);
```
Comment (rui-mo) [inline cpp/velox/compute/RowConstructor.cc:58]: Could you update the null count as well?

File: cpp/velox/compute/RowConstructor.cc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto argsCopy = args;
+
+    BufferPtr nulls = AlignedBuffer::allocate<char>(bits::nbytes(rows.size()), context.pool());
```
Comment (rui-mo) [inline cpp/velox/compute/RowConstructor.cc:36]: Based on below Velox's implementation, how about setting the default value and removing the bits setting?

```
template <>
inline BufferPtr AlignedBuffer::allocate<bool>(
    size_t numElements,
    velox::memory::MemoryPool* pool,
    const std::optional<bool>& initValue) {
```

File: cpp/velox/compute/RowConstructor.cc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    auto* nullsPtr = nulls->asMutable<uint64_t>();
+    for (size_t i = 0; i < rows.size(); ++i) {
+      bits::setBit(nullsPtr, i);
+    }
```
Comment (rui-mo) [inline cpp/velox/compute/RowConstructor.cc:40]: Can be removed if default value is set.

**Comment by rui-mo** [comment]
Please also fix the code format issue.

File: cpp/velox/compute/RowConstructor.cc (hunk 1 above)
Comment (lviiii) [reply cpp/velox/compute/RowConstructor.cc]: Done

File: cpp/velox/compute/RowConstructor.cc (hunk 2 above)
Comment (lviiii) [reply cpp/velox/compute/RowConstructor.cc:36]: Done

File: cpp/velox/compute/RowConstructor.cc (hunk 3 above)
Comment (lviiii) [reply cpp/velox/compute/RowConstructor.cc]: Done

File: cpp/velox/compute/RowConstructor.cc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    auto argsCopy = args;
+
+    BufferPtr nulls = AlignedBuffer::allocate<char>(bits::nbytes(rows.size()), context.pool());
+    auto* nullsPtr = nulls->asMutable<uint64_t>();
+    for (size_t c = 0; c < argsCopy.size(); c++) {
+      auto arg = argsCopy[c].get();
+      if (arg->mayHaveNulls()) {
+        rows.applyToSelected([&](vector_size_t i) {
+          if (arg->isNullAt(i)) {
+            bits::setNull(nullsPtr, i, true);
```
Comment (rui-mo) [inline cpp/velox/compute/RowConstructor.cc:43]: Why not updating the null count here to save the for loop?

Comment (lviiii) [reply cpp/velox/compute/RowConstructor.cc:43]: If we do that, null count is not correct in loop.

**Comment by lviiii** [comment]
> 

Done

File: cpp/velox/compute/RowConstructor.cc [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto argsCopy = args;
+
+    BufferPtr nulls = AlignedBuffer::allocate<char>(
+        bits::nbytes(rows.size()), context.pool(), 1);
+    auto* nullsPtr = nulls->asMutable<uint64_t>();
+    auto cntNull = 0;
+    rows.applyToSelected([&](vector_size_t i) {
+      bits::clearNull(nullsPtr, i);
```
Comment (rui-mo) [inline cpp/velox/compute/RowConstructor.cc:40]: Thanks for iterating on this. Still have two comments. Why this is needed? I guess the default value is already set.

File: cpp/velox/compute/RowConstructor.cc [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+          if (arg->mayHaveNulls()) {
+            if (arg->isNullAt(i)) {
+              bits::setNull(nullsPtr, i, true);
+              cntNull++;
+              break;
+            }
+          }
+        }
+      }
+    });
```
Comment (rui-mo) [inline cpp/velox/compute/RowConstructor.cc:53]: How about sth like below, would it work?
```
    rows.applyToSelected([&](vector_size_t i) {
      if (!bits::isBitNull(nullsPtr, i)) {
        for (size_t c = 0; c < argsCopy.size(); c++) {
          auto arg = argsCopy[c].get();
          if (arg->mayHaveNulls() && arg->isNullAt(i)) {
            bits::setNull(nullsPtr, i, true);
            cntNull++;
            break;
          }
        }
      }
    });
```

**Review by rui-mo — APPROVED** [review]

File: cpp/velox/jni/jni_wrapper.cc [hunk 7]
```diff
@@ -127,6 +128,7 @@ Java_io_glutenproject_vectorized_ExpressionEvaluatorJniWrapper_nativeDoValidate(
   std::unique_ptr<core::ExecCtx> execCtx_ =
       std::make_unique<core::ExecCtx>(pool.get(), queryCtx_.get());
 
+  velox::compute::registerAllFunctions();
```
Comment (jinchengchenghh) [inline cpp/velox/jni/jni_wrapper.cc:131]: Why do you add this? I received this coredump
```
#8  __GI___libc_malloc (bytes=87) at malloc.c:3051
#9  0x00007fdd594bcb39 in operator new(unsigned long) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#10 0x00007fdd59557d3d in void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007fd93ca83ecc in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct_aux<char const*> (
    this=0x7fdacf4f1620, __beg=0x7fd9440fbb68 "min(38, max(a_precision - a_scale, b_precision - b_scale) + max(a_scale, b_scale) + 1)",
    __end=0x7fd9440fbbbe "") at /usr/include/c++/9/bits/basic_string.h:251
#12 0x00007fd93ca7d723 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (
    this=0x7fdacf4f1620, __beg=0x7fd9440fbb68 "min(38, max(a_precision - a_scale, b_precision - b_scale) + max(a_scale, b_scale) + 1)",
    __end=0x7fd9440fbbbe "") at /usr/include/c++/9/bits/basic_string.h:270
#13 0x00007fd93ca7729b in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> > (this=0x7fdacf4f1620, __s=0x7fd9440fbb68 "min(38, max(a_precision - a_scale, b_precision - b_scale) + max(a_scale, b_scale) + 1)",
    __a=...) at /usr/include/c++/9/bits/basic_string.h:531
#14 0x00007fd93feeb3f9 in std::_Optional_payload_base<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::_Storage<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, false>::_Storage<char const (&) [87]> (
    this=0x7fdacf4f1620) at /usr/include/c++/9/optional:223
#15 0x00007fd93feeb472 in std::_Optional_payload_base<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::_Optional_payload_base<char const (&) [87]> (this=0x7fdacf4f1620, __tag=...) at /usr/include/c++/9/optional:115
#16 0x00007fd93feeb035 in std::_Optional_payload<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, true, false, false>::_Optional_payload_base<char const (&) [87]> (this=0x7fdacf4f1620) at /usr/include/c++/9/optional:356
#17 0x00007fd93feeb05f in std::_Optional_payload<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, false, false, false>::_Optional_payload_base<char const (&) [87]> (this=0x7fdacf4f1620) at /usr/include/c++/9/optional:390
#18 0x00007fd93feeb092 in std::_Optional_base<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, false, false>::_Optional_base<char const (&) [87], false> (this=0x7fdacf4f1620) at /usr/include/c++/9/optional:482
#19 0x00007fd93feeacfa in std::optional<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::optional<char const (&) [87], true> (this=0x7fdacf4f1620, __t=...) at /usr/include/c++/9/optional:699
#20 0x00007fd93fea39fc in facebook::velox::functions::(anonymous namespace)::decimalAddSubtractSignature ()
    at /mnt/DP_disk1/code/gluten/tools/build/velox_ep/velox/functions/prestosql/DecimalArithmetic.cpp:280
#21 0x00007fd93fea4860 in facebook::velox::functions::registerVectorFunction_udf_decimal_add (name="plus")
    at /mnt/DP_disk1/code/gluten/tools/build/velox_ep/velox/functions/prestosql/DecimalArithmetic.cpp:362
#22 0x00007fd93df76644 in facebook::velox::functions::registerArithmeticFunctions ()
    at /mnt/DP_disk1/code/gluten/tools/build/velox_ep/velox/functions/prestosql/registration/ArithmeticFunctionsRegistration.cpp:96
--Type <RET> for more, q to quit, c to continue without paging--
#23 0x00007fd93df6f10f in facebook::velox::functions::prestosql::registerArithmeticFunctions ()
    at /mnt/DP_disk1/code/gluten/tools/build/velox_ep/velox/functions/prestosql/registration/RegistrationFunctions.cpp:37
#24 0x00007fd93df6f1cf in facebook::velox::functions::prestosql::registerAllScalarFunctions ()
    at /mnt/DP_disk1/code/gluten/tools/build/velox_ep/velox/functions/prestosql/registration/RegistrationFunctions.cpp:85
#25 0x00007fd93caa27c0 in velox::compute::registerAllFunctions ()
    at /mnt/DP_disk1/code/gluten/cpp/velox/compute/RegistrationAllFunctions.cc:34
#26 0x00007fd93ca6c906 in Java_io_glutenproject_vectorized_ExpressionEvaluatorJniWrapper_nativeDoValidate (env=0x7fdd5637ca58,
    obj=0x7fdacf4f1a50, planArray=0x7fdacf4f1a48) at /mnt/DP_disk1/code/gluten/cpp/velox/jni/jni_wrapper.cc:131

```

Comment (lviiii) [reply cpp/velox/jni/jni_wrapper.cc:131]: These functions were registered at Velox validation, but it caused Gluten's functions were overwrote. So I put function registration in JNI.

**Pull request merged**